### PR TITLE
Support `nerdctl run --gpus`

### PR DESCRIPTION
**Pull request by ktock**
Fixes: https://github.com/containerd/nerdctl/issues/248

This PR adds `--gpus` option to `nerdctl run` based on containerd's GPU support (`github.com/containerd/containerd/contrib/nvidia` by https://github.com/containerd/containerd/pull/2330).

```
# nerdctl run --gpus all --rm -it nvidia/cuda:9.0-base nvidia-smi
```

For compose (https://github.com/compose-spec/compose-spec/blob/master/deploy.md#devices):

```yaml
version: "3.8"
services:
  demo:
    image: nvidia/cuda:9.0-base
    command: nvidia-smi
    deploy:
      resources:
        reservations:
          devices:
          - capabilities: ["utility"]
            driver: nvidia
            count: all
```

`nvidia-container-cli` is needed.

### CODE REVIEW

**Comment by AkihiroSuda** [comment]
:+1:

Can we support Compose as well? https://github.com/compose-spec/compose-spec/blob/master/deploy.md#devices

Can be another PR.

File: README.md [hunk 1]
```diff
@@ -307,6 +307,9 @@ Metadata flags:
 Shared memory flags:
 - :whale: `--shm-size`: Size of `/dev/shm`
 
+GPU flags:
+- :whale: `--gpus`: GPU devices to add to the container ('all' to pass all GPUs). `nvidia-container-cli` is needed.
```
Comment (AkihiroSuda) [inline README.md:311]: Can we have `./docs/gpu.md` to explain all the options?

We should also clarify how to set up GPU for rootless. (Can be another PR)

**Comment by ktock** [comment]
Added compose support and docs.

File: docs/gpu.md [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+
+## Options for `nerdctl run --gpus`
+
+`nerdctl run --gpus` is compatible to [`docker run --gpus`](https://docs.docker.com/engine/reference/commandline/run/#access-an-nvidia-gpu).
+
+You can specify number of GPUs to use via `--gpus` option.
+The following example exposes all available GPUs.
+
+```
+nerdctl run -it --rm --gpus all ubuntu:20.04 nvidia-smi
```
Comment (AkihiroSuda) [inline docs/gpu.md:20]: `ubuntu:20.04` -> `nvidia/cuda:9.0-base ` might be more useful?

File: docs/gpu.md [hunk 3]
```diff
@@ -0,0 +1,61 @@
+# Using GPUs inside containers
+
+nerdctl provides docker-compatible NVIDIA GPU support.
+
+## Prerequisites
+
+- NVIDIA Drivers
+  - Same requirement as when you use GPUs on Docker. For details, please refer to [the doc by NVIDIA](https://docs.nvidia.com/datacenter/cloud-native/container-toolkit/install-guide.html#pre-requisites).
+- `nvidia-container-cli`
+  - containerd relies on this CLI for setting up GPUs inside container. You can install this via [`libnvidia-container` package](https://docs.nvidia.com/datacenter/cloud-native/container-toolkit/arch-overview.html#libnvidia-container).
```
Comment (AkihiroSuda) [inline docs/gpu.md:10]: Did you try rootless (on cgroup v1)?

I guess it needs setting `no-cgroups = true`
https://github.com/moby/moby/issues/38729#issuecomment-463493866

**Comment by AkihiroSuda** [comment]
I'll release nedctl v0.9.0 after merging this.

File: pkg/composer/serviceparser/serviceparser.go [hunk 4]
```diff
@@ -193,6 +215,60 @@ func getMemLimit(svc compose.ServiceConfig) (types.UnitBytes, error) {
...
+			if dev.Driver != "" {
+				e = append(e, fmt.Sprintf("driver=%s", dev.Driver))
+			}
+			if len(dev.IDs) > 0 {
+				e = append(e, fmt.Sprintf("device=%s", strings.Join(dev.IDs, ",")))
+			}
+			if dev.Count != 0 {
+				e = append(e, fmt.Sprintf("count=%d", dev.Count))
+			}
+
```
Comment (fahedouch) [inline pkg/composer/serviceparser/serviceparser.go:256]: `count` and  `device_ids` are mutually exclusive. we should define one field at a time. is it configured somewhere ?Start a reviewAdd single commentCancel

File: docs/gpu.md (hunk 3 above)
Comment (ktock) [reply docs/gpu.md:10]: It doesn't work as of now.

We might need to patch `github.com/containerd/containerd/contrib/nvidia` for allowing to pass `--no-cgroup` option to `nvidia-container-cli`. 

Containerd doesn't use `nvidia-container-runtime` (instead, it executes `nvidia-container-cli` directly) so we cannot use `/etc/nvidia-container-runtime/config.toml` for nerdctl.

File: pkg/composer/serviceparser/serviceparser.go (hunk 4 above)
Comment (ktock) [reply pkg/composer/serviceparser/serviceparser.go:256]: This is check at https://github.com/containerd/nerdctl/pull/251/files#diff-9d2133247dbc7d059776b8e61aaeb758cde413729d8ed7eb45ba1febbdb43ca6R139-R141

File: docs/gpu.md (hunk 2 above)
Comment (ktock) [reply docs/gpu.md]: Fixed this.

File: docs/gpu.md (hunk 3 above)
Comment (ktock) [reply docs/gpu.md:10]: A *very hacky* workaround for this is to wrap `nvidia-container-cli` to forcefully specify `--no-cgroups`.

```
mkdir -p /opt/nvidia/bin
mv /usr/bin/nvidia-container-cli /opt/nvidia/bin/
cat <<'EOF' > /usr/bin/nvidia-container-cli
#!/bin/bash
/opt/nvidia/bin/nvidia-container-cli ${@:1:($#-1)} --no-cgroups ${@:$#}
EOF
```

Comment (AkihiroSuda) [reply docs/gpu.md:10]: Opened https://github.com/containerd/containerd/issues/5603 for discussion

**Comment by fahedouch** [comment]
> I'll release nedctl v0.9.0 after merging this.

I will clear this ticket tonight https://github.com/containerd/nerdctl/pull/239 . It will be  good to have it in 0.9 :)

File: docs/gpu.md (hunk 3 above)
Comment (ktock) [reply docs/gpu.md:10]: @AkihiroSuda https://github.com/containerd/containerd/pull/5604 is merged.
Updated this PR to use `--no-cgroup` and now it works in rootless environment as well.

`replace` directive is needed in `go.mod` to forcefully point to the latest commit of containerd.

File: go.mod [hunk 5]
```diff
@@ -27,13 +28,15 @@ require (
...
 	github.com/sirupsen/logrus v1.8.1
 	github.com/urfave/cli/v2 v2.3.0
 	golang.org/x/sync v0.0.0-20210220032951-036812b2e83c
-	golang.org/x/sys v0.0.0-20210420072515-93ed5bcd2bfe
+	golang.org/x/sys v0.0.0-20210426230700-d19ff857e887
 	golang.org/x/term v0.0.0-20210406210042-72f3dc4e9b72
 	gotest.tools/v3 v3.0.3
 )
+
+replace github.com/containerd/containerd => github.com/containerd/containerd v1.5.1-0.20210614183500-0a3a77bc4453
```
Comment (AkihiroSuda) [inline go.mod:42]: Why replace?

Comment (ktock) [reply go.mod:42]: Without replace, `go mod tidy` wants to point to `v1.5.2`.

Comment (AkihiroSuda) [reply go.mod:42]: 😞

**Review by AkihiroSuda — APPROVED** [review]
Thanks
